### ISSUE TITLE
Bad performance on QUIC connection

### ISSUE THREAD

**Issue by maskit** [apache/trafficserver#3552]
Current performance (throughput) on QUIC connections is terrible.

**Comment by maskit** [comment]
One of causes is here.
https://github.com/apache/trafficserver/blob/b23d1c57c03b689a3b06b2b7719e9674bd9dfb8b/iocore/net/QUICNetVConnection.cc#L1585-L1588

Performance got much better with small (2ms) WRITE_READY_INTERVAL, it was still 10x slower than H2 though.
This event scheduling will be removed once we introduce connect+bind.

**Comment by maskit** [comment]
```diff
diff --git a/iocore/net/QUICNetVConnection.cc b/iocore/net/QUICNetVConnection.cc
index a0285bda2..e0657ada8 100644
--- a/iocore/net/QUICNetVConnection.cc
+++ b/iocore/net/QUICNetVConnection.cc
@@ -65,7 +65,7 @@ static constexpr uint32_t MAX_PACKET_OVERHEAD       = 62; // Max long header len
 static constexpr uint32_t MAX_STREAM_FRAME_OVERHEAD = 24;
 // static constexpr uint32_t MAX_CRYPTO_FRAME_OVERHEAD   = 16;
 static constexpr uint32_t MINIMUM_INITIAL_PACKET_SIZE = 1200;
-static constexpr ink_hrtime WRITE_READY_INTERVAL      = HRTIME_MSECONDS(20);
+static constexpr ink_hrtime WRITE_READY_INTERVAL      = HRTIME_MSECONDS(2);
 static constexpr uint32_t PACKET_PER_EVENT            = 32;
 static constexpr uint32_t MAX_CONSECUTIVE_STREAMS     = 8; //< Interrupt sending STREAM frames to send ACK frame
 
diff --git a/iocore/net/UnixUDPNet.cc b/iocore/net/UnixUDPNet.cc
index 2f291165b..77c506f93 100644
--- a/iocore/net/UnixUDPNet.cc
+++ b/iocore/net/UnixUDPNet.cc
@@ -796,6 +796,7 @@ UDPQueue::SendPackets()
   bytesThisSlot = INT_MAX;
 
 sendPackets:
+  fprintf(stderr, "----- sendPackets: %llu\n", now / HRTIME_MSECOND);
   sentOne       = false;
   bytesThisPipe = (int32_t)bytesThisSlot;
 
@@ -896,6 +897,7 @@ UDPQueue::send(UDPPacket *p)
 {
   // XXX: maybe fastpath for immediate send?
   outQueue.push((UDPPacketInternal *)p);
+  fprintf(stderr, "------------ send: %llu\n", Thread::get_hrtime_updated() / HRTIME_MSECOND);
 }
 
 #undef LINK

```

```
------------ send: 1542676010653
------------ send: 1542676010653
------------ send: 1542676010653
----- sendPackets: 1542676010694
----- sendPackets: 1542676010697
------------ send: 1542676010699
------------ send: 1542676010699
------------ send: 1542676010699
```

**Comment by scw00** [comment]
send side should be the same as receive side. The ET_UDP is always blocked with epoll for a several ms and then check the send queue. So it should be a signal to terminate the blocking.

**Comment by scw00** [comment]
Also it should be changed on master branch.

**Comment by maskit** [comment]
Tested with #4628. It may fixed some cases, but I still see the delay. Total transfer time (about 42 second for 128MB) didn't change.

```
------------ send: 1542679189811
------------ send: 1542679189811
----- sendPackets: 1542679189815
----- sendPackets: 1542679189818
------------ send: 1542679189818
------------ send: 1542679189818
```

```
------------ send: 1542679189832
------------ send: 1542679189832
----- sendPackets: 1542679189883
----- sendPackets: 1542679189885
------------ send: 1542679189886
------------ send: 1542679189886
```

**Comment by maskit** [comment]
> Also it should be changed on master branch.

I don't remember the change on master. Should I also merge master?

**Comment by scw00** [comment]
The send side has not implement the signalActivity in ET_UDP.

**Comment by maskit** [comment]
Average delay after `send` was about 20ms.
There are 1500 sendPackets log and usually they show up 3 times in a row.
So total delay is about 20ms * 500 = 10,000ms.

Since total transfer time is currently 42 sec, there should be some more slow places.

**Comment by maskit** [comment]
Pushed d64f8c6e96961915de13a3e197ae50bba90aa4f7

Changed a couple of constant values.
- WRITE_READY_INTERVAL was simply too long. (It will be removed when we implement connect+bind)
- PACKET_PER_EVENT was was too small. 3-4 events were needed even if data was available.

**Comment by maskit** [comment]
With two changes above total transfer time is now 35 sec.

**Comment by maskit** [comment]
With #4629 (and a few changes on QUIC side), it's 4.3 sec !!
Surprisingly, it's faster than H1 over TLS (6.5sec).

**Comment by maskit** [comment]
I merged #4629 and pushed 880e9f6732e2ed09706b0d504684924647b3287c. Without any debug prints, now it's 2.8 sec.

Although we cannot simply compare the result with other protocol versions because clients are different, the result is unbelievable. It's faster than non-TLS HTTP/1.1.

|Protocol| Time for 128MB transfer (sec) |
|--|--|
| 0.9 over QUIC | 2.8 |
| 2 over TLS | 6.7 |
| 1.1 over TLS | 6.5 |
| 1.1 non-TLS | 4.3 |

Command lines:
```
$time ./examples/client -q --timeout=1 -d GET128m  127.0.0.1 4433 
$ curl --http1.1 -k -v -w 'total=%{time_total}' https://localhost:8443/128m
$ curl -k -v -w 'total=%{time_total}' https://localhost:8443/128m
```
*QUIC client is ngtcp2

There would be places we can improve but I'm closing this because it's not terrible now. 

@scw00 Thank you for the great work!
